Airside_Functions.cpp

```diff
@@ -0,0 +1,458 @@
+        {
+            return MAVLINK_ENCODING_FAIL;
+        }
+    }
+
+    if (start_index > 1)
+    {
+        for (int i = 0; i<2;i++)
+        {
+            message_buffer[message_len-2+i] = ptr_in_byte[start_index-2+i]; // load the last 2 checksum bytes
+            //printf("copying byte: %d / %d   |   current byte : %hhx\n", message_len-2+i, message_len, message_buffer[message_len-2+i]);
+        }
+        memcpy(message, message_buffer, message_len);
+
+        return MAVLINK_ENCODING_OKAY;
+    }
+    else
+    {
+        return MAVLINK_ENCODING_FAIL;
+    }
+}
+
+//---------------------------------------- tests -----------------------------------------------------------------------------
+//an example of how to use the encoder and decoder
+
+int test__encode_then_decode(void)
+{
+   PIGO_GPS_LANDING_SPOT_t landing_spot = {
+       1,
+       2,
+       3,
+       4,
+   };
+   POGI_GPS_t GPS_Test = {
+       1,2,3,
+   };
+    PIGO_WAYPOINTS_t warg_GPS = 
+    {
+        2, //int32_t latitude;
+        3, //int32_t longitude;
+        4, //int32_t altitude;
+        5, //int32_t turnpoint;
+        9,
+    };
+
+    PIGO_GIMBAL_t gimbal_cmd = 
+    {
+        2, //pitch
+        1, //yaw
+    };
+    POGI_Euler_Angle_t angle_cmd = {
+        1, //yaw
+        2, //pitch
+        3, //roll
+    };
+    uint16_t inte = 8;
+    one_byte_uint_cmd_t uint8_cmd = 
+    {
+        8,
+    };
+
+    single_bool_cmd_t bool_cmd = 
+    {
+        1,
+    };
+
+
+    mavlink_message_t encoded_msg;
+    memset(&encoded_msg, 0, sizeof(mavlink_message_t));
+    //printf("space holder here\n");
+    uint8_t encoderStatus = Mavlink_airside_encoder(MESSAGE_ID_ERROR_CODE, &encoded_msg, (const uint8_t*) &uint8_cmd);
+    //uint8_t encoderStatus = Mavlink_airside_encoder(MESSAGE_ID_GPS, &encoded_msg, (const uint8_t*) &landing_spot);
+    //uint8_t encoderStatus = Mavlink_airside_encoder(MESSAGE_ID_EULER_ANGLE_CAM, &encoded_msg, (const uint8_t*) &angle_cmd);
+
+    if (encoderStatus == MAVLINK_ENCODING_FAIL)
+    {
+        return 0;
+    }
+
+    //---------------------------------- decoding starts ---------------------------------- 
+
+    mavlink_decoding_status_t decoderStatus = MAVLINK_DECODING_INCOMPLETE;
+
+    char decoded_message_buffer[50]; //256 is the max payload length
+
+    // the following few lines imitates how a decoder is used when it gets one byte at a time from a serial port
+    unsigned char* ptr_in_byte = (unsigned char *) &encoded_msg;
+    PIGO_Message_IDs_e message_type = MESSAGE_ID_NONE;
+
+    for( int i = 0; i < 50; i++) // 50 is just a random number larger than message length (for GPS message length is 39)
+    {
+        if (decoderStatus != MAVLINK_DECODING_OKAY)
+        {
+            //printf("copying byte: %d  |  current byte : %hhx\n", i, ptr_in_byte[i]);
+            decoderStatus = Mavlink_airside_decoder(&message_type, ptr_in_byte[i], (uint8_t*) &decoded_message_buffer);
+        }
```

Comment:
Say the airside decoder returns MAVLINK_DECODING_OKAY the first time around. Won't that mean it will not be executed again in the for loop? 

I think I am missing something here

Reply:
Ya I think the part you missed is how the decoder works. :) 

the decoder is called after each byte is received, there's a parse functions gets called in the decoder, and it can accumulate bytes and return message_received  once it checks the CRC and everything, then it proceed to the actual decoding message part, where if it's a valid message, the status will be set to OKAY.

So the status will get changed to OKAY only if the entire message is received. All other occasions would be either stilling getting incoming bytes or has an error.